#library(tidyverse)
#library(viridis)
library(ape)
library(devtools)
#install_github("nclark-lab/RERconverge")
library(phangorn)
library(RERconverge)
#library(data.table)
#library(gt)
#library(ggrepel)
#library(ggpubr)
#library(stringr)


#### -------------------------####
# Test whether alternate species tree generated by gene tree discordance analysis produce the same tree.
rerpath = find.package('RERconverge')
#read in tree file
########Produce your trees using phangorn
estimatePhangornTreeAll(alndir = '/home/lyusuf/scratch/FISH_WGA/Genomes/WGA_CYPRINODON/MafFilter/AA_alignments_TDG/Done_Alignments/Fasta_AA_alignments', treefile = 'FocalBranchGoodeid_AltTopo1.nwk', output.file = 'AltTopologyGoodeid1', format='fasta', type='AA')
estimatePhangornTreeAll(alndir = '/home/lyusuf/scratch/FISH_WGA/Genomes/WGA_CYPRINODON/MafFilter/AA_alignments_TDG/Done_Alignments/Fasta_AA_alignments', treefile = 'FocalBranchPoecilid_AltTopo1.nwk', output.file = 'AltTopologyPoecilid1', format='fasta', type='AA')

#name phangorn trees
#Treesconverge_ALTG1=readTrees(paste('AltTopologyGoodeid1'), max.read = 17000)
#Treesconverge_ALTP1=readTrees(paste('AltTopologyPoecilid1'), max.read = 17000)

# 
# master_ALTG1<-Treesconverge_ALTG1$masterTree
# #master_ALTG2<-Treesconverge_ALTG2$masterTree
# master_ALTP1<-Treesconverge_ALTP1$masterTree
# #master_ALTP2<-Treesconverge_ALTP2$masterTree

#spp_names_ALTG1<-master_ALTG1$tip.label
#spp_names_ALTG2<-master_ALTG2$tip.label
#spp_names_ALTP1<-master_ALTP1$tip.label
#spp_names_ALTP2<-master_ALTP2$tip.label
#Now get evolutionary rates that will be corrected for heteroscedasity.
#fishviviparity_ALTG1= getAllResiduals(Treesconverge_ALTG1, useSpecies = spp_names_ALTG1, transform = "sqrt", weighted = T, scale = T)#
#fishviviparity_ALTP1= getAllResiduals(Treesconverge_ALTP1, useSpecies = spp_names_ALTP1 transform = "sqrt", weighted = T, scale = T)#

#Produce a trait tree for each plausible discordant topology and re-do the analysis.
#G1
trait_tree_G1=read.tree("FocalBranchGoodeid_AltTopo1_TRAIT.nwk")
phenotype_viviparity_G1=tree2Paths(trait_tree_G1, Treesconverge_ALTG1)
corViviparity_G1=correlateWithBinaryPhenotype(fishviviparity_ALTG1, phenotype_viviparity_G1, min.sp=21, min.pos=2)
Significant_correlation_G1 <- corViviparity_G1 %>% filter(P < 0.05)
write.table(Significant_correlation_G1, file = "Revisions_ALTTopologyG1_RerSignificantResults.txt", sep='\t', row.names = T, quote=F)

#P1
trait_tree_P1=read.tree("FocalBranchPoecilid_AltTopo1_TRAIT.nwk")
phenotype_viviparity_P1=tree2Paths(trait_tree_P1, Treesconverge_ALTP1)
corViviparity_P1=correlateWithBinaryPhenotype(fishviviparity_ALTP1, phenotype_viviparity_P1, min.sp=21, min.pos=2)
Significant_correlation_P1 <- corViviparity_P1 %>% filter(P < 0.05)
write.table(Significant_correlation_P1, file = "Revisions_ALTTopologyP1_RerSignificantResults.txt", sep='\t', row.names = T, quote=F)


